### ISSUE TITLE
Standalone not working with Microchip cryptoauthlib

### ISSUE THREAD

**Issue by seanpmaddox** [Nuitka/Nuitka#1106]
Greetings,

We're utilizing Microchip's cryptoauthlib (https://github.com/MicrochipTech/cryptoauthlib) which is a C package for which Microchip has provided a python ctypes layer (https://github.com/MicrochipTech/cryptoauthlib/tree/main/python) to enable using the library from python.

Modules in their python library rely on using the handle returned from a call to ctypes.cdll.LoadLibrary() to load cryptoauthlib.so which you'll find in the load_cryptoauthlib() function of their library.py module  (https://github.com/MicrochipTech/cryptoauthlib/blob/main/python/cryptoauthlib/library.py).  I've pasted the code from this function and one it depends on below.

def _force_local_library():
    """
    In some environments loading seems to fail under all circumstances unless
    brute forcing it.
    """
    curr_path = os.path.dirname(__file__)

    if sys.platform.startswith('win'):
        return os.path.join(curr_path, "cryptoauth.dll")
    elif sys.platform.startswith('darwin'):
        return os.path.join(curr_path, "libcryptoauth.dylib")
    else:
        return os.path.join(curr_path, "libcryptoauth.so")

def load_cryptoauthlib(lib=None):
    """
    Load CryptoAauthLib into Python environment
    raise LibraryLoadError if cryptoauthlib library can't be loaded
    """
    global _CRYPTO_LIB      # pylint: disable=global-statement
    if lib is not None:
        _CRYPTO_LIB = lib
    else:
        try:
            os.environ['PATH'] = os.path.dirname(__file__) + os.pathsep + os.environ['PATH']
            library_file = find_library('cryptoauth')
            if library_file is None:
                library_file = _force_local_library()
            _CRYPTO_LIB = cdll.LoadLibrary(library_file)
        except:
            raise LibraryLoadError('Unable to find cryptoauthlib. You may need to reinstall')

Once I've compiled to a single file via "sudo python3 -m nuitka --onefile --show-scons --plugin-enable=pylint-warnings my_python_main.py" and then run the resulting binary I get the following error;

Traceback (most recent call last):
  File "/tmp/.mount_qastlegs2hws/cryptoauthlib/library.py", line 102, in load_cryptoauthlib
  File "/tmp/.mount_qastlegs2hws/ctypes/__init__.py", line 434, in LoadLibrary
  File "/tmp/.mount_qastlegs2hws/ctypes/__init__.py", line 356, in __init__
OSError: /tmp/.mount_qastlegs2hws/cryptoauthlib/libcryptoauth.so: cannot open shared object file: No such file or directory

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/tmp/.mount_qastlegs2hws/qastle.py", line 33, in <module>
    from qastle_COMMON import QASTLE_SW_NAME as qastleSW
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
  File "/tmp/.mount_qastlegs2hws/qastle_COMMON.py", line 23, in <module qastle_COMMON>
    import qastle_ATECCX08 as crypto
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
  File "/tmp/.mount_qastlegs2hws/qastle_ATECCX08.py", line 104, in <module qastle_ATECCX08>
    import cryptoauthlib as crypto
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
  File "/tmp/.mount_qastlegs2hws/cryptoauthlib/__init__.py", line 20, in <module cryptoauthlib>
  File "/tmp/.mount_qastlegs2hws/cryptoauthlib/library.py", line 104, in load_cryptoauthlib
cryptoauthlib.exceptions.LibraryLoadError: Unable to find cryptoauthlib. You may need to reinstall

I looked but have not found a method I should use to resolve the issue.

Nuitka version, python version and platform;
0.6.15
Python: 3.7.3 (default, Jul 25 2020, 13:03:44)
Executable: /usr/bin/python3
OS: Linux
Arch: armv7l

Nuitka was installed via pip3 install -U nuitka
I installed python3, i believe, via sudo apt-get install 

To fully reproduce I believe you would need to install the Microchip cryptoauthlib though I suspect given the issue you may already know of a method that can be used to provide the handle relative to the nuitka compiled version OR provide guidance to adjust the library calls so that they still leverage cryptoauthlib.so but in a different manner that works with nuitka.

thx - sean

**Comment by kayhayen** [comment]
It seems this is for commercial usage, maybe you are interested in this: https://nuitka.net/pages/commercial.html

The DLL indeed needs a special treatment that can be added in a similar fashion than others in the implicit imports plugin, you can have a look or as a customer, I would do it for you.

**Comment by seanpmaddox** [comment]
Hi Kay - yes, it is for commercial use.  At the moment we are exploring different options to compile and exploring whether or not we can fully compile and run using Nuitka.  If it proves out that we can and this looks to be the best route for us we will purchase a subscription.  I've only been exploring Nuitka for a few days.

From what I can tell though commercial doesn't actually resolve this issue correct?

I did come up with what seems to be a working solution/work around;

1. use --include-plugin-directory=/usr/local/lib/python3.7/dist-packages/cryptoauthlib
2. change their load_cryptoauthlib() function to _CRYPTO_LIB = cdll.LoadLibrary('/usr/local/lib/python3.7/dist-packages/cryptoauthlib/libcryptoauth.so') making it a full path reference and remove the other file finding code they had

With these changes I do not get an error and I have been able to, so far, utilize their library to interface with their chips on our board.  what i will very likely do is create a separate directory into which any/all SO's like this will be placed and change the --include and cdll.LoadLibrary() calls to use this directory so that I am not pulling in unnecessary things including source code.

I will also take a look at the implicit imports plugins.

Much appreciate the insight.

thx - sean
